immutable/src/main/java/org/jdbi/v3/core/mapper/reflect/ImmutableMapper.java

```diff
@@ -0,0 +1,308 @@
+                "Mapping immutable type %s could not match properties for columns: %s",
+                type.getSimpleName(),
+                unmatchedColumns));
+        }
+        return result;
+    }
+
+    private RowMapper<T> specialize0(ResultSet rs,
+                                     StatementContext ctx,
+                                     List<String> columnNames,
+                                     List<ColumnNameMatcher> columnNameMatchers,
+                                     List<String> unmatchedColumns) {
+        final List<RowMapper<?>> mappers = new ArrayList<>();
+        final List<PropertyDescriptor> properties = new ArrayList<>();
+        final List<PropertyDescriptor> unmatchedDescriptors = new ArrayList<>(info);
+
+        for (PropertyDescriptor descriptor : info) {
+            Nested anno = Stream.of(descriptor.getReadMethod(), descriptor.getWriteMethod())
+                .filter(Objects::nonNull)
+                .map(m -> m.getAnnotation(Nested.class))
+                .filter(Objects::nonNull)
+                .findFirst()
+                .orElse(null);
+
+            if (anno == null) {
+                String paramName = prefix + paramName(descriptor);
+
+                findColumnIndex(paramName, columnNames, columnNameMatchers, () -> debugName(descriptor))
+                    .ifPresent(index -> {
+                        Type type = descriptor.getReadMethod().getGenericReturnType();
+                        ColumnMapper<?> mapper = ctx.findColumnMapperFor(type)
+                            .orElse((r, n, c) -> r.getObject(n));
+
+                        mappers.add(new SingleColumnMapper<>(mapper, index + 1));
+                        properties.add(descriptor);
+
+                        unmatchedColumns.remove(columnNames.get(index));
+                        unmatchedDescriptors.remove(descriptor);
+                    });
+            } else {
+                String nestedPrefix = prefix + anno.value();
+
+                RowMapper<?> nestedMapper = nestedMappers
+                    .computeIfAbsent(descriptor, d -> new ImmutableMapper(d.getPropertyType(), nestedPrefix))
+                    .specialize0(rs, ctx, columnNames, columnNameMatchers, unmatchedColumns);
+
+                mappers.add(nestedMapper);
+                properties.add(descriptor);
+                unmatchedDescriptors.remove(descriptor);
+            }
+        }
+
+        if (mappers.isEmpty() && columnNames.size() > 0) {
+            throw new MappingException(String.format("Mapping immutable type %s " +
+                "didn't find any matching columns in result set", type));
+        }
+
+        return (r, c) -> {
+            final Map<Method, Object> invocations = new HashMap<>();
+            for (int i = 0; i < mappers.size(); i++) {
+                RowMapper<?> mapper = mappers.get(i);
+                PropertyDescriptor property = properties.get(i);
+                Object value = mapper.map(r, ctx);
```

Comment:
Would be good to wrap the raw values in unmodifiable wrappers if they are a map, set, or collection.

Later on we could also think about supporting array properties.